### PR TITLE
Add support for xthin blocks.

### PR DESCRIPTION
**Pull request by dagurval**
Keeps support for bloom thin blocks. Allows parallel fething of both
types for same block.

### CODE REVIEW

**Comment by dagurval** [comment]
This is an incomplete and untested implementation of xthin blocks. What's notably lacking is:
- Receiving xthin blocks.
- Supporting "CXRequestThinBlockTx", although I'm not sure I want to.

The end goal is supporting both types of thin blocks in parallel, avoid cluttering main.cpp and having good unittest coverage.

I created this PR even though I'm not finished to show progress. I will be busy next two weeks but will continue after that.

**Comment by dagurval** [comment]
I just pushed further work of xthinblocks.

* It's well tested with unit tests, which all pass for me.
* RPC testing stage 1 passes
* I have not tried stage 2 yet.
* I have not tried connecting to BU nodes.

As far as I know all of xthin protocol has been implemented. I will do some testing now.

**Comment by dagurval** [comment]
I have been running this PR on two nodes on mainnet for a day now without issues.

Here is a log snippet I find interesting. It shows what this PR is capable of: This is a block downloaded in parallel with a _bloom thin block_ from an XT node **and** two _xthin blocks_ from BU nodes:

```
2016-03-23 22:06:34 requesting 0000000000000000060ade823ab6b76b0458f304709c570446da2f1c0b9e5fb3 from peer 20 (1 of 3 parallel)
2016-03-23 22:06:34 sending: get_xthin (3235 bytes) peer=20
2016-03-23 22:06:34 got inv: block 0000000000000000060ade823ab6b76b0458f304709c570446da2f1c0b9e5fb3  new peer=14
2016-03-23 22:06:34 requesting 0000000000000000060ade823ab6b76b0458f304709c570446da2f1c0b9e5fb3 from peer 14 (2 of 3 parallel)
2016-03-23 22:06:34 sending: get_xthin (3235 bytes) peer=14
2016-03-23 22:06:35 got inv: block 0000000000000000060ade823ab6b76b0458f304709c570446da2f1c0b9e5fb3  new peer=417
2016-03-23 22:06:35 requesting 0000000000000000060ade823ab6b76b0458f304709c570446da2f1c0b9e5fb3 from peer 417 (3 of 3 parallel)
2016-03-23 22:06:35 received: merkleblock (32080 bytes) peer=417
2016-03-23 22:06:35 received stub for block 0000000000000000060ade823ab6b76b0458f304709c570446da2f1c0b9e5fb3 (peer 417) 64 out of 992 txs missing
2016-03-23 22:06:35 received: xthinblock (133159 bytes) peer=14
2016-03-23 22:06:35 received xthinblock 0000000000000000060ade823ab6b76b0458f304709c570446da2f1c0b9e5fb3 from peer=14
2016-03-23 22:06:35 reassembled thin block for 0000000000000000060ade823ab6b76b0458f304709c570446da2f1c0b9e5fb3 (584937 bytes)
2016-03-23 22:06:35 Received block 0000000000000000060ade823ab6b76b0458f304709c570446da2f1c0b9e5fb3 0.36 seconds after requesting it from peer 20
2016-03-23 22:06:35 Received block 0000000000000000060ade823ab6b76b0458f304709c570446da2f1c0b9e5fb3 0.17 seconds after requesting it from peer 14
2016-03-23 22:06:35 Received block 0000000000000000060ade823ab6b76b0458f304709c570446da2f1c0b9e5fb3 0.08 seconds after requesting it from peer 417
2016-03-23 22:06:35 UpdateTip: new best=0000000000000000060ade823ab6b76b0458f304709c570446da2f1c0b9e5fb3  height=403981  log2_work=84.353766  tx=118021894  date=2016-03-23 22:05:43 progress=1.000000  cache=0.3MiB(0tx)
2016-03-23 22:06:35 got inv: block 0000000000000000060ade823ab6b76b0458f304709c570446da2f1c0b9e5fb3  have peer=19
2016-03-23 22:06:35 received: xthinblock (133159 bytes) peer=20
2016-03-23 22:06:35 received xthinblock 0000000000000000060ade823ab6b76b0458f304709c570446da2f1c0b9e5fb3 from peer=20
2016-03-23 22:06:35 already had block 0000000000000000060ade823ab6b76b0458f304709c570446da2f1c0b9e5fb3, ignoring xthinblock peer=20
2016-03-23 22:06:35 ignoring xthinblock 0000000000000000060ade823ab6b76b0458f304709c570446da2f1c0b9e5fb3 that peer is not working on peer=20
2016-03-23 22:06:35 got inv: block 0000000000000000060ade823ab6b76b0458f304709c570446da2f1c0b9e5fb3  have peer=13
2016-03-23 22:06:35 received: get_xthin (6885 bytes) peer=18
2016-03-23 22:06:35 sending: xthinblock (8743 bytes) peer=18
2016-03-23 22:06:35 received: get_xthin (7002 bytes) peer=10
2016-03-23 22:06:35 sending: xthinblock (8743 bytes) peer=10
2016-03-23 22:06:35 received: get_xthin (11915 bytes) peer=16
2016-03-23 22:06:35 sending: xthinblock (8743 bytes) peer=16
2016-03-23 22:06:36 received: get_xthin (9284 bytes) peer=5
2016-03-23 22:06:36 sending: xthinblock (8743 bytes) peer=5
```

**Comment by dgenr8** [comment]
That's impressive.  Three peers provided a 585KB block in .36s or less, and the chain was updated in ~1s.

Was peer 417 on your local network? ;)

**Comment by dagurval** [comment]
Not bad! :-). I don't know if it's in the same data center or not, log is pruned so I can't tell. I'm using a $20 droplet from Digital Ocean.

The node is restarted. Doing some more testing - the node is connected to 14 Bitcoin Unlimited nodes and 7 XT0.11.0E nodes, in addition to some other random incoming nodes. It's configured to not to download full blocks:

```
2016-03-24 12:35:32 Received block 000000000000000003d5342d100984b20a60c2ffa6feda493465f588eabf60bf 0.19 seconds after requesting it from peer 19
2016-03-24 12:37:50 Received block 00000000000000000194ff157daa78234a01cc258e8dd51229612fa68b935e3f 0.31 seconds after requesting it from peer 13
2016-03-24 12:45:10 Received block 00000000000000000315ba23bf1858cd774559fd8ce507085d7106043023233d 0.35 seconds after requesting it from peer 18
2016-03-24 12:49:34 Received block 000000000000000004042a82c8dfa7f82251068d7c74e8fa6986257706a5d4a7 0.78 seconds after requesting it from peer 118
2016-03-24 12:51:20 Received block 0000000000000000065b4f76ff809bc91c9b335c530b46706f5c6dbba0f39c0f 0.15 seconds after requesting it from peer 118
2016-03-24 13:03:40 Received block 000000000000000004a49ec6a44a7e56a4b895b28f00701a93a000000cc6b0b9 0.44 seconds after requesting it from peer 118
2016-03-24 13:05:42 Received block 0000000000000000019009b19b950b9dea644f0008e7a5001c56a26ec6e4d426 0.05 seconds after requesting it from peer 19
2016-03-24 13:20:32 Received block 000000000000000000d9293caa86d7fbdff14949113bc06be43e3be81cff9c97 0.52 seconds after requesting it from peer 293
2016-03-24 13:27:40 Received block 00000000000000000166f680002c5787d06a2fa5cb687ee4bca8907cd0795933 0.20 seconds after requesting it from peer 13
2016-03-24 13:28:15 Received block 00000000000000000217907ef69c986feab010c44004c504f30eadc852d34f32 0.02 seconds after requesting it from peer 19
2016-03-24 13:29:49 Received block 0000000000000000039612a8989124ffd55dad06c00cde6717e019955d274740 0.05 seconds after requesting it from peer 19
2016-03-24 13:34:28 Received block 0000000000000000005a2d44d7aeeb5bd0ddf93242e8c35dee60a7bd624fb2db 0.24 seconds after requesting it from peer 13
2016-03-24 13:43:03 Received block 000000000000000003771c53f3850c8b83dc87df35c38fd7472d86e4d2208964 0.36 seconds after requesting it from peer 293
2016-03-24 13:45:54 Received block 000000000000000003745bec1040678aae0b2941f73c5aa7e41bdf216f1ae3bc 0.08 seconds after requesting it from peer 19
2016-03-24 14:21:44 Received block 0000000000000000018914d196c226a0f8c38fa209feaa364de7ad3bc2e69036 1.23 seconds after requesting it from peer 293
2016-03-24 14:33:35 Received block 0000000000000000057437da4b3bcc0ec981ab7b55092d5af80029e6a9e2e4ba 1.06 seconds after requesting it from peer 293
2016-03-24 14:36:16 Received block 000000000000000000984db96d6cf8dc29ea92a8ec88987e741ba00ac4f49406 0.33 seconds after requesting it from peer 13
2016-03-24 14:38:38 Received block 00000000000000000228171ce1599aa8bb5ae93e7a7c7366fd0e195517db8fc4 0.25 seconds after requesting it from peer 13
2016-03-24 14:52:44 Received block 000000000000000004740090b8fad58dbf02a554214464afbb96ed7f085276c2 0.63 seconds after requesting it from peer 293
2016-03-24 15:00:38 Received block 000000000000000003e2ea2d7bd738ba48eb8dd2b5bfc7690c9578a4871624fc 0.33 seconds after requesting it from peer 19
2016-03-24 15:09:32 Received block 0000000000000000044ff00079a752a81c78f593aaaca0b86d0dc69a5f6f932d 0.56 seconds after requesting it from peer 293
2016-03-24 15:30:18 Received block 000000000000000005184f37880b0cb81223df43110511d03c6bb8e74997f4fd 0.91 seconds after requesting it from peer 293
2016-03-24 15:40:49 Received block 000000000000000004d7731b6c3f03fcd612f7f314d7c041378765d6ff6347cf 0.77 seconds after requesting it from peer 19
2016-03-24 15:45:39 Received block 0000000000000000050b9f4c185f2bd1266751d7b5809c94e2d6bd6a4651abcf 2.52 seconds after requesting it from peer 1510
2016-03-24 15:55:31 Received block 000000000000000005281900d7cede67a4c02cd48f5338bed8f7e2ca85c9965d 0.60 seconds after requesting it from peer 1510
2016-03-24 16:08:20 Received block 000000000000000000e9d7104f3ae88cc7c01c641f703048c0953ac6ee3ff09d 0.95 seconds after requesting it from peer 1510
2016-03-24 16:19:53 Received block 0000000000000000036272adca5fa54418edd99b3ddc39b298ac149ccad6692b 0.54 seconds after requesting it from peer 1510
2016-03-24 16:29:11 Received block 0000000000000000038accbb2a7adfe005dbd0e3c5c1b457a05f11dd0168c663 0.76 seconds after requesting it from peer 1510
2016-03-24 16:32:56 Received block 0000000000000000003785c7ad119c4ac8b2948f82901f8066ca39606b3a0a66 0.82 seconds after requesting it from peer 1510
2016-03-24 16:54:05 Received block 00000000000000000610b3615aac9954062fd80eed5683f14f4fb126da9ac4f4 1.00 seconds after requesting it from peer 1510
2016-03-24 16:59:06 Received block 0000000000000000017f4f7bec594b0ea5052e4b364b46ab599aa2baec95a909 0.72 seconds after requesting it from peer 1510
2016-03-24 17:02:15 Received block 0000000000000000045970843f11408a532eaf4a0e91cc2872bcb610efcfc720 1.21 seconds after requesting it from peer 1510
2016-03-24 17:02:25 Received block 000000000000000005fbe3a8acb14e5482ac5bcf55a995222bc09f6ec49738e8 0.18 seconds after requesting it from peer 1510
2016-03-24 17:03:46 Received block 000000000000000003f6f0fb33d54ec49bd3096ec14cf2dd1373ffb24435638a 0.31 seconds after requesting it from peer 1510
2016-03-24 17:19:31 Received block 00000000000000000667f9adca1019ebe4395f0053b172b6a157e5ec4f5f5870 0.96 seconds after requesting it from peer 1510
2016-03-24 17:20:23 Received block 000000000000000001f2d6471dfc25ad93f68c677f7a22bdae73899e0258236a 0.16 seconds after requesting it from peer 1510
2016-03-24 17:36:37 Received block 00000000000000000683889b59994864c7805fc20588ca9abb3b9429967062a8 1.33 seconds after requesting it from peer 1510
2016-03-24 17:54:21 Received block 00000000000000000464bea06e2ff5c55b3afa9f030722500a36b38546cac951 1.82 seconds after requesting it from peer 1510
2016-03-24 17:56:21 Received block 0000000000000000013341c2549c57a7e0a211a441253989eae10a5cf9ddf534 1.74 seconds after requesting it from peer 1510
```
That's an **average of 0.65** and **median of 0.54** :-).

I've seen this error a few times now, I guess I've messed up a LogPrint or LogPrintf somewhere:
```
************************
EXCEPTION: St13runtime_error
tinyformat: Not enough conversion specifiers in format string
bitcoin in ProcessMessages()
```

**Comment by ptschip** [comment]
i've been seeing some xthins coming in from one of your XT nodes to one
of my BU nodes - nice :)

2016-03-25 00:57:29 Received thinblock
0000000000000000055cbedb7c5a7d18c125e07f6c552249ef4ad28c474f78d3 from
peer 188.166.108.147:50636 (45). Size 18098 bytes.
2016-03-25 00:57:29 Received block
0000000000000000055cbedb7c5a7d18c125e07f6c552249ef4ad28c474f78d3 in 0.59
seconds

**Comment by dagurval** [comment]
@ptschip That is one of my nodes :-). I obviously used BU as reference when implementing so I will add credits to you in the final commit squash! 

I did not understand the code that seeded the bloom filter when requesting a thin block, so this PR has a different (, a too) simple implementation of that. When this is merged, could you submit a PR with better bloom filter seeding?

**Comment by ptschip** [comment]
for the seeded bloom filter was it the linear decay algoritm that didn't
make sense "nBloomDecay"?...sorry we probably should have had a little
more documentation in there.  Basically that is just for gradually
decreasing the proportion of the bloom filter elements as the number of
elements increases.  It ends up increasing the false positive rate as
the mempool increases in size, however, because the mempool now has most
of the transactions we can "lean" on the mempool in a probabilistic way
and end up still not having to do a re-request.  You don't need to do
that but it does help to keep the bloom filter size down and save some
bandwidth.

sounds like you're almost there...

**Comment by dagurval** [comment]
I have fixed the logprint errors. This PR has been running fine on my nodes and I don't know of any blocking issues, so I have removed the WIP tag.

**Comment by dagurval** [comment]
I discovered a problem during initial block sync when running extended tests. I have pushed a commit with a fix for the issue.

All extended tests complete for me (except 1, I skipped prune.py).

File: src/protocol.h [hunk 1]
```diff
@@ -79,7 +79,12 @@ enum {
     // NODE_BLOOM means the node is capable and willing to handle bloom-filtered connections.
     // All nodes with protocol version < 70011 support bloom-filter without advertising this bit.
     // Core nodes with protocol >= 70011 do not support bloom-filter unless this bit is advertised.
-    NODE_BLOOM = (1 << 2)
+    NODE_BLOOM = (1 << 2),
+
+    // FIXME: Remove or rename. This one will likely be used for segwit.
```
Comment (rebroad) [inline src/protocol.h]: is being used for SegWit

**Pull request merged**